source/geod24/concurrency.d

```diff
@@ -1446,3 +1502,665 @@ private:
+{
+    struct HyperLoopMessage
+    {
+        int id;
+        MonoTime time;
+    }
+
+    FiberScheduler scheduler = new FiberScheduler;
+    auto chn1 = new Channel!HyperLoopMessage();
+    auto chn2 = new Channel!HyperLoopMessage();
+    auto chn3 = new Channel!HyperLoopMessage();
+
+    scheduler.spawn(
+        () {
+            HyperLoopMessage msg;
+
+            for (int i = 0; i < 1000; ++i)
+            {
+                assert(chn2.read(msg));
+                assert(msg.id % 3 == 1);
+                msg.id++;
+                msg.time = MonoTime.currTime;
+                assert(chn3.write(msg));
+            }
+        }
+    );
+
+    scheduler.spawn(
+        () {
+            HyperLoopMessage msg;
+
+            for (int i = 0; i < 1000; ++i)
+            {
+                assert(chn1.read(msg));
+                assert(msg.id % 3 == 0);
+                msg.id++;
+                msg.time = MonoTime.currTime;
+                assert(chn2.write(msg));
+            }
+        }
+    );
+
+    scheduler.start(
+        () {
+            HyperLoopMessage msg = {
+                id : 0,
+                time : MonoTime.currTime
+            };
+
+            for (int i = 0; i < 1000; ++i)
+            {
+                assert(chn1.write(msg));
+                assert(chn3.read(msg));
+                assert(msg.id % 3 == 2);
+                msg.id++;
+                msg.time = MonoTime.currTime;
+            }
+        }
+    );
+}
+
+// Multiple writer threads writing to a buffered channel
+// Reader should receive all messages
+@system unittest
+{
+    FiberScheduler scheduler = new FiberScheduler;
+    immutable int n = 5000;
+    auto chn1 = new Channel!int(n/10);
+
+    shared int sharedVal = 0;
+    shared int writer_sum = 0;
+
+    auto t1 = new InfoThread({
+        FiberScheduler scheduler = new FiberScheduler();
+        scheduler.start(
+            () {
+                int val = atomicOp!"+="(sharedVal, 1);
+                while (chn1.write(val))
+                {
+                    atomicOp!"+="(writer_sum, val);
+                    val = atomicOp!"+="(sharedVal, 1);
+                }
+            }
+        );
+    });
+    t1.start();
+
+    auto t2 = new InfoThread({
+        FiberScheduler scheduler = new FiberScheduler();
+        scheduler.start(
+            () {
+                int val = atomicOp!"+="(sharedVal, 1);
+                while (chn1.write(val))
+                {
+                    atomicOp!"+="(writer_sum, val);
+                    val = atomicOp!"+="(sharedVal, 1);
+                }
+            }
+        );
+    });
+    t2.start();
+
+    scheduler.start(
+        () {
+            int reader_sum, readVal, count;
+
+            while(chn1.read(readVal))
+            {
+                reader_sum += readVal;
+                if (count++ == n) chn1.close();
+            }
+
+            thread_joinAll();
+            assert(reader_sum == writer_sum);
+        }
+    );
+}
```

Comment:
I think you'll need to change your text editor settings. Needs trailing newline. :>